### ISSUE TITLE
Mixed Precision Training on Lesson 3 Planet

### ISSUE THREAD

**Issue by tydlwav** [fastai/fastai#1903]
**Describe the bug**
I have been trying to use to_fp16 on the [planet notebook](https://github.com/fastai/course-v3/blob/master/nbs/dl1/lesson3-planet.ipynb). My 8GB RTX 2070 cannot fit the default fp32 mode with the default batch size of 64. It seems to me that fp16 could be a better option than reducing batch size.

I've had some experience working with fp16 (using Nvidia apex AMP), so I applied the same methodology here. Unfortunately, I was not able to get fp16 to train properly. I've tried to simply add .to_fp16(), use .to_fp16(dynamic=True), .to_fp16(loss_scale=128), .to_fp16(loss_scale=1024). None of them would make the model train.

**Provide your installation details**

```text
=== Software === 
python        : 3.7.3
fastai        : 1.0.50.post1
fastprogress  : 0.1.20
torch         : 1.0.1.post2
nvidia driver : 418.43
torch cuda    : 10.0.130 / is available
torch cudnn   : 7402 / is enabled

=== Hardware === 
nvidia gpus   : 1
torch devices : 1
  - gpu0      : 7949MB | GeForce RTX 2070

=== Environment === 
platform      : Linux-4.19.32-1-MANJARO-x86_64-with-arch-Manjaro-Linux
distro        : #1 SMP PREEMPT Wed Mar 27 18:55:07 UTC 2019
conda env     : fastai
python        : /home/michael/miniconda3/envs/fastai/bin/python
sys.path      : 
/home/michael/miniconda3/envs/fastai/lib/python37.zip
/home/michael/miniconda3/envs/fastai/lib/python3.7
/home/michael/miniconda3/envs/fastai/lib/python3.7/lib-dynload
/home/michael/miniconda3/envs/fastai/lib/python3.7/site-packages
```

**To Reproduce**
Run the [planet notebook](https://github.com/fastai/course-v3/blob/master/nbs/dl1/lesson3-planet.ipynb) with `.to_fp16()`. 

**Expected behavior**
F score should be around 0.92 to 0.93, but with fp16 training, it is only 0.47 to 0.48. 

**Additional context**
I doubt that this could be an issue with the fastai library rather than an implementation issue of the course code (or that I'm using `to_fp16()` incorrectly). Just to be safe, I'm opening issues on both the course repo and here. 
Details on the loss graph is [here](https://github.com/fastai/course-v3/issues/266).

**Comment by sgugger** [comment]
I tried to reproduce by adding `to_fp16()` in the cell that defined the learner:
```
acc_02 = partial(accuracy_thresh, thresh=0.2)
f_score = partial(fbeta, thresh=0.2)
learn = cnn_learner(data, arch, metrics=[acc_02, f_score]).to_fp16()
```
and it worked without any issue. Please provide the exact code you are running as we can't fix a bug we aren't able to reproduce.

**Comment by tydlwav** [comment]
That is exactly what I did. I will upload a .py file once I run it as a script.

**Comment by tydlwav** [comment]
Code I ran (on both dev build from master and 1.0.50.post1): 
```
from fastai.vision import *

path = Config.data_path()/'planet'
path.mkdir(parents=True, exist_ok=True)

tfms = get_transforms(flip_vert=True, max_lighting=0.1, max_zoom=1.05, max_warp=0.)

np.random.seed(42)
src = (ImageList.from_csv(path, 'train_v2.csv', folder='train-jpg', suffix='.jpg')
       .split_by_rand_pct(0.2)
       .label_from_df(label_delim=' '))

data = (src.transform(tfms, size=128)
        .databunch().normalize(imagenet_stats))

arch = models.resnet50

acc_02 = partial(accuracy_thresh, thresh=0.2)
f_score = partial(fbeta, thresh=0.2)
learn = cnn_learner(data, arch, metrics=[acc_02, f_score]).to_fp16()
# learn = cnn_learner(data, arch, metrics=[acc_02, f_score])

lr = 0.01

learn.fit_one_cycle(5, slice(lr))
```

Results for fp16:
```
epoch     train_loss  valid_loss  accuracy_thresh  fbeta     time    
0         0.842592    0.754226    0.187966         0.476119  00:46                   
1         0.840402    0.750478    0.190125         0.477969  00:45                   
2         0.840895    0.753238    0.189376         0.475770  00:46                   
3         0.839311    0.745062    0.189790         0.477772  00:46                   
4         0.843568    0.753511    0.189231         0.477805  00:46                   
Total time: 03:50
```

Results for fp32:
```
epoch     train_loss  valid_loss  accuracy_thresh  fbeta     time    
0         0.128140    0.110074    0.943771         0.904181  00:42                   
1         0.108145    0.098180    0.949904         0.912403  00:42                   
2         0.101861    0.092210    0.951466         0.921294  00:44                   
3         0.095462    0.087542    0.956465         0.924710  00:41                   
4         0.091450    0.086625    0.955811         0.924713  00:43                   
Total time: 03:33
```

**Comment by sgugger** [comment]
Might be a hardware/drive/cuda/cudnn problem. Does mixed precision work in one setting? Or does it always fail?

**Comment by tydlwav** [comment]
I've run the [cifar notebook](https://github.com/fastai/fastai/blob/master/examples/cifar.ipynb) and it works fine. I ran fp16 with pytorch and apex on CUDA 10 to train cifar10 on resnet34 from scratch in February without any issues, so I doubt it is a hardware/driver/cuda/cudnn thing.

**Comment by tydlwav** [comment]
Interesting, the issue is magically solved if I don't run `lr_find()` for fp16. Once I run `lr_find()`, the model doesn't train. I wonder how `lr_find` would affect the training. Could be something to do with [this callback function](https://github.com/fastai/fastai/blob/master/fastai/callbacks/lr_finder.py#L36)?

**Comment by sgugger** [comment]
Ah, that *is* different from the code you sent me ;)
Able to reproduce now, this is linked to the learning rate finder test somehow.

**Comment by tydlwav** [comment]
Yeah, was a little messy downloading the notebook as `.py` and selecting parts of it. Apparently, I didn't copy over the lr_find line. The notebook was always reproducible, though.

**Comment by sgugger** [comment]
Should be fixed now.
